### ISSUE TITLE
Pinned z3 version == 4.8.8.0 makes it hard to use the fuzzingbook as a library

### ISSUE THREAD

**Issue by rindPHI** [uds-se/fuzzingbook#115]
Hi all,

I'm using the fuzzingbook packages as a library for my research prototypes. I'm facing the problem that the pinned z3 version apparently required by the concolic fuzzer, 4.8.8.0, is not working with my tool, which needs something like 4.8.13.0. Would it be possible to relax the "==" constraint on z3-solver to a ">=" constraint? I could install 4.8.13.0 anyway in my venv and ignore the warning, but then my tox tests fail (I didn't find a --no-deps option for tox).

Thanks!

Best,
Dominic

**Comment by michaelmera** [comment]
I don't know if the constraint in the fuzzingbook can/should be relaxed, but in the meantime a fix for you to make tox work is to add `pip install z3-solver==4.8.13` as the first command in your tox.ini `commands` variables.

Note also that if you don't recreate your venv (don't `tox -r`/`tox --recreate`), dependencies should not be re-installed, which mean you can also install it manually once.

**Comment by rindPHI** [comment]
Thanks so much, this indeed solves my problem! It's still not optimal that one gets a warning during the installation of requirements, but I can very much live with that I guess. Maybe it would be nice to modularize the fuzzingbook packages, such that one can decide not to import the concolic fuzzer and there's no strict requirement of z3 involved. But that's a separate issue ;)

**Comment by andreas-zeller** [comment]
What we know is that recent z3 versions have changed their output format, which now lacks info we need in `ConcolicFuzzer.ipynb` and `SymbolicFuzzer.ipynb`. I'll be happy to relax the z3 version range if you can 

1. determine the latest version that still works with the above notebooks; or
2. adjust the two notebooks to work with the latest z3 version.

In the meantime, maybe using a virtual environment for `fuzzingbook` (and the specific z3 version) may help.

**Comment by andreas-zeller** [comment]
Confirming that the highest z3 version `ConcolicFuzzer` and `SymbolicFuzzer` work with is 4.8.9.0. @vrthra and @rindPHI – feel free to adapt these notebooks to recent z3 versions.

**Comment by vrthra** [comment]
[Tracking bug in Z3](https://github.com/Z3Prover/z3/issues/5763) before we can update.

**Comment by vrthra** [comment]
See commit b195516d996a1f0f17e554af7d6bd16b3800bc07 Unfortunately, 4.8.10.0 is the maximum version that ConcolicFuzzer will work with because the new versions break with bitvec comparison to a string.

**Comment by vrthra** [comment]
[Tracking bug in z3](https://github.com/Z3Prover/z3/issues/5764) for bitvec comparison. The current working version is 4.8.10.0 (assuming https://github.com/Z3Prover/z3/issues/5763 gets fixed)

**Comment by rindPHI** [comment]
Obviously, in early 2021 z3 changed the internal representation of Strings to "the string theory custom type instead of bit-vectors" (-> https://github.com/Z3Prover/z3/discussions/4939#discussioncomment-305212) instead of bit vectors, which is why the code in your commit message (https://github.com/uds-se/fuzzingbook/commit/b195516d996a1f0f17e554af7d6bd16b3800bc07) stopped working.

Why exactly do you need to work with bit vectors?

**Comment by vrthra** [comment]
I need to implement `ord()` for a single character. If you can suggest a different means, I can incorporate that (or you can check that in).

**Comment by andreas-zeller** [comment]
Hi folks – thanks for your work! I plan to release a new pip fuzzingbook package (1.0.5) today or tomorrow; if we could get compatibility with the newest Z3 before that, that would be great.

**Comment by rindPHI** [comment]
z3 supports str.to_code / str.from_code (https://github.com/Z3Prover/z3/issues/5007), but not via the Python API. What you can do, however, is

```python
import z3


def z3_ord(s: z3.SeqRef) -> z3.ArithRef:
    return z3.parse_smt2_string("(assert (= 42 (str.to_code s)))", decls={"s": s})[0].children()[1]


f = z3_ord(z3.String("s"))
print(f)
# str.to_code(s)
print(f.sort())
# Int
```

Would this help you?

**Comment by rindPHI** [comment]
Another example (tested, works):

```python
s = z3.Solver()
v = z3.String("v")
s.add(z3.IntVal(ord("D")) == z3_ord(v))
s.check()
assert s.model()[v].as_string() == "D"
```

**Comment by rindPHI** [comment]
Here another example including conversion to the other direction, with hardened z3_ord / z3_chr functions that accept arbitrary expressions (of the correct types) as arguments:

```python
from typing import Union, Optional, Dict, Generator, Set

import z3


def visit_z3_expr(
        e: Union[z3.ExprRef, z3.QuantifierRef],
        seen: Optional[Dict[Union[z3.ExprRef, z3.QuantifierRef], bool]] = None) -> \
        Generator[Union[z3.ExprRef, z3.QuantifierRef], None, None]:
    if seen is None:
        seen = {}
    elif e in seen:
        return

    seen[e] = True
    yield e

    if z3.is_app(e):
        for ch in e.children():
            for e in visit_z3_expr(ch, seen):
                yield e
        return

    if z3.is_quantifier(e):
        for e in visit_z3_expr(e.body(), seen):
            yield e
        return


def is_z3_var(e: z3.ExprRef) -> bool:
    return z3.is_const(e) and e.decl().kind() == z3.Z3_OP_UNINTERPRETED


def get_all_vars(e: z3.ExprRef) -> Set[z3.ExprRef]:
    return {sub for sub in visit_z3_expr(e) if is_z3_var(sub)}


def z3_ord(str_expr: z3.SeqRef) -> z3.ArithRef:
    return z3.parse_smt2_string(
        f"(assert (= 42 (str.to_code {str_expr.sexpr()})))",
        decls={str(c): c for c in get_all_vars(str_expr)}
    )[0].children()[1]


def z3_chr(int_expr: z3.ArithRef) -> z3.SeqRef:
    return z3.parse_smt2_string(
        f"(assert (= \"4\" (str.from_code {int_expr.sexpr()})))",
        decls={str(c): c for c in get_all_vars(int_expr)}
    )[0].children()[1]


s = z3.Solver()
v = z3.String("v")
s.add(z3.StringVal("D") == z3_chr(z3_ord(z3.StringVal("D"))))
print(s.check())
# sat
```

**Comment by rindPHI** [comment]
@vrthra I hope that helps you. Since I'm not familiar with the implementation of your concolic executor, it would be great if you could integrate this workaround yourself ;) If there are things that cannot be solved with my ideas, I'm of course happy to look for other solutions or extend the one above.

**Comment by rindPHI** [comment]
Another example:

```python
# Lots of strings starting with "D"
prev: Set[str] = set()
for _ in range(100):
    s = z3.Solver()
    v = z3.String("v")

    s.add(z3_ord(z3.SubString(v, 0, 1)) == z3.IntVal(ord("D")))
    for prev_sol in prev:
        s.add(z3.Not(v == z3.StringVal(prev_sol)))

    assert s.check() == z3.sat
    m = s.model()
    sol = m[v].as_string()
    print(sol)
    prev.add(sol)
```

What does *not* work is `s.add(z3_ord(v[0]) == z3.IntVal(ord("D")))` instead of the `z3.SubString` call, since the character access returns a Char/Unicode value and not a String, and there, `to_code` is not defined. Thus the use of `SubString`, which is no limitation though.

**Comment by vrthra** [comment]
> z3 supports str.to_code / str.from_code ([Z3Prover/z3#5007](https://github.com/Z3Prover/z3/issues/5007)), but not via the Python API. What you can do, however, is
> 
> ```python
> import z3
> 
> 
> def z3_ord(s: z3.SeqRef) -> z3.ArithRef:
>     return z3.parse_smt2_string("(assert (= 42 (str.to_code s)))", decls={"s": s})[0].children()[1]
> 
> 
> f = z3_ord(z3.String("s"))
> print(f)
> # str.to_code(s)
> print(f.sort())
> # Int
> ```
> 
> Would this help you?

Can you also provide `z3_chr()` in the above style? Given an int, convert to  string?

**Comment by vrthra** [comment]
```
$ cat x.smt
(echo "starting Z3...")
(declare-const a Int)
(assert (= (str.from_code a) "A"))
(check-sat)
(get-model)
$ ./4.8.14.0/bin/z3 smt.string_solver=z3str3 x.smt 
starting Z3...
sat
(
  (define-fun a () Int
    0)
)
$ cat y.smt                                       
(echo "starting Z3...")
(declare-const s String)
(assert (= (str.to_code s) 65))
(check-sat)
(get-model)

$ ./4.8.14.0/bin/z3 smt.string_solver=z3str3 y.smt
starting Z3...
sat
(
  (define-fun s () String
    "A")
)
```
and

```
$ ./4.8.14.0/bin/z3  x.smt                        
starting Z3...
sat
(
  (define-fun a () Int
    65)
)
```

**Comment by vrthra** [comment]
Unfortunately, removing `z3str3` results in multiple smt evaluations returning `Gaveup`.

**Comment by rindPHI** [comment]
My experience was also to avoid z3str3, the standard seq solver is more stable.

So you get timeouts w/o z3str3 also using 4.8.14? For me, this z3 version works much better (fewer timeouts) than 4.8.8. Do you have some sample representative constraints that time out?

**Comment by andreas-zeller** [comment]
Fixed in 1.0.6 and later
